#  Exploratory Data Analysis(EDA)

## 1. Getting to know our dataset


In [1]:
#Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as datetime
import warnings
warnings.filterwarnings('ignore')

C:\Users\tsed\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.4
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
#Loading the dataset 
mpg_df=pd.read_csv('D:\\Tsed\\Python\\Projects\\Mile Per Gallon\\Dataset\\mpg.csv')
mpg_df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa                 chevy s-10  

[398 rows x 9 columns]

####  INSIGHTS
    :The dataset contains information about various vehicle models and their fuel efficiency(mpg) along with other attribute
    :It has 398 rows & 9 columns. The index, the columns and the rows look formal.

The dataset typically includes the following attributes:

MPG: Miles per gallon 
Cylinders:Number of cylinders in the engine
Displacement: total volume displaced by all the cylinders(TDC-BDC)
Horsepower: power output of Engine
Weight: Vehicle weight
Acceleration: how quickly a vehicle can increase its speed.
Model Year: Year of the vehicle model 
Origin: country of the vehicle where it is manufactured
Name: Name of the car

### 1.1 Quick overview of mpg

In [3]:
# Quick overview of the dataset
mpg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


####  INSIGHTS
    : model_year's data type is int which has to be in datetime type. 
    : Weight's data type should be float.
    : Column horsepower has 6 missing values.
    : the data contains 4 float, 3 int & 2 object datatypes.

### 1.2. Exploring variables

In [4]:
# Identifying categorical variables
cat_df=mpg_df.select_dtypes('object')
cat_cols=cat_df.columns
cat_df.head(3)

origin                       name
0    usa  chevrolet chevelle malibu
1    usa          buick skylark 320
2    usa         plymouth satellite

In [5]:
# validating mpg variable as being numerical
for i in cat_df.columns:
    print(i,':', cat_df.value_counts(i).count())

origin : 3
name : 305


#### INSIGHTS
    : In the origin column there is 3 categories
    : Name column has 305 different names of cars

In [6]:
# Identifying numerical vaiables
num_df=mpg_df.select_dtypes('number')
num_cols=num_df.columns
num_df.head(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  model_year
0  18.0          8         307.0       130.0    3504          12.0          70
1  15.0          8         350.0       165.0    3693          11.5          70
2  18.0          8         318.0       150.0    3436          11.0          70

In [7]:
# validating mpg variable as being numerical
for i in num_df.columns:
    print(i,':', num_df.value_counts(i).count())

mpg : 129
cylinders : 5
displacement : 82
horsepower : 93
weight : 351
acceleration : 95
model_year : 13


In [8]:
#Validating should cyliders be numerical?
mpg_df['cylinders'].value_counts().sort_values(ascending=False)

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

In [9]:
cat_df['cylinders']=mpg_df['cylinders']
cat_df.head(3)

origin                       name  cylinders
0    usa  chevrolet chevelle malibu          8
1    usa          buick skylark 320          8
2    usa         plymouth satellite          8

In [10]:
mpg_df['origin'].value_counts().sort_values(ascending=False)

usa       249
japan      79
europe     70
Name: origin, dtype: int64

In [11]:
# Remove cylinders from numerical dataframe
num_df.drop('cylinders',axis=1,inplace=True)
num_df.head(3)

mpg  displacement  horsepower  weight  acceleration  model_year
0  18.0         307.0       130.0    3504          12.0          70
1  15.0         350.0       165.0    3693          11.5          70
2  18.0         318.0       150.0    3436          11.0          70

#### INSIGHTS
    :It shows the dataset has 5 kind of cylinders 3,4,5,6,8. we can't do mathematical operation like sum,subtract or so 
     and get meaningful result. Therefore lets add clinders to categorical varaibles. 
    :So we have 3 categorical and  6 numerical columns.
    :Most of the vehicle in our dataset are taken from usa. most of the vehicle in our dataset have 4 cylinders.

### 2.1. Summary of descirptive statistics: For only numerical variables

In [12]:
# Print the summary statistics
num_stat_df=num_df.describe()

In [13]:
#append the median in the summary stat dataframe
new_num_stat=num_stat_df.append(num_df.median(),ignore_index='False')
new_num_stat.index=['count', 'mean', 'std', 'min', '25%', '50%', '75%', 'max','medain']
new_num_stat

mpg  displacement  horsepower       weight  acceleration  \
count   398.000000    398.000000  392.000000   398.000000    398.000000   
mean     23.514573    193.425879  104.469388  2970.424623     15.568090   
std       7.815984    104.269838   38.491160   846.841774      2.757689   
min       9.000000     68.000000   46.000000  1613.000000      8.000000   
25%      17.500000    104.250000   75.000000  2223.750000     13.825000   
50%      23.000000    148.500000   93.500000  2803.500000     15.500000   
75%      29.000000    262.000000  126.000000  3608.000000     17.175000   
max      46.600000    455.000000  230.000000  5140.000000     24.800000   
medain   23.000000    148.500000   93.500000  2803.500000     15.500000   

        model_year  
count   398.000000  
mean     76.010050  
std       3.697627  
min      70.000000  
25%      73.000000  
50%      76.000000  
75%      79.000000  
max      82.000000  
medain   76.000000

#### INSIGHTS
    :Again we can see horsepower has 6 missing values from the count row. 
    :Outliers: 
    :mpg: mean>median------------- slight difference, tends to normal distribution
    :displacement: mean>median-----higher d/ce & right skewed
    :horsepower: mean>median------ higher d/ce & right skewed
    :weight: mean>median-----------higher d/ce & right skewed
    :acceleration: mean & median---almost equal & seems normal distribution
    :model_year:mean=meidan-------almost equal & normal distribution

let's make sure our observations are right? and What created the right-skewed distribution?